jax/_src/lax/parallel.py

```diff
@@ -347,42 +353,41 @@ def _allreduce_translation_rule(prim, c, val, *, axis_name, axis_index_groups,
-# and associative over axis names if they support tuples. That is,
-# they have to satisfy:
-#   collective(x, ('i', 'j')) == collective(x, ('j', 'i'))
-#                             == collective(collective(x, 'j'), 'i')
-def _split_axis_comm_assoc(primitive, split_name, args, params):
-  axis_names = params['axis_name']
-  assert isinstance(axis_names, tuple)
-  if params['axis_index_groups'] is not None:
-    raise NotImplementedError("axis_index_groups not supported in axis splitting. "
-                              "Please open a feature request!")
-  remaining_axes = list(axis_names)
-  remaining_axes.remove(split_name)
-  remaining_axes = tuple(remaining_axes)
-  split_params = dict(params, axis_name=split_name)
-  remain_params = dict(params, axis_name=remaining_axes)
-  split_result = primitive.bind(*args, **split_params)
-  if not primitive.multiple_results:
-    split_result = (split_result,)
-  return primitive.bind(*split_result, **remain_params)
-
-# NB: This is only used for collectives that do not include the vmapped axis name,
-#     which is why the rule is so simple. All other collectives go through split_axis.
+
+# This is only used for collectives that do not include the vmapped axis name,
+# which is why the rule is so simple.
 def _collective_batcher(prim, args, dims, **params):
   return prim.bind(*args, **params), dims if prim.multiple_results else dims[0]
 
-def _batched_reduction_collective(prim, if_mapped, if_unmapped,
-                                  vals_in, dims_in, axis_size,
-                                  axis_name, axis_index_groups):
+def _batched_reduction_collective(
+    prim, if_mapped, if_unmapped, frame, vals_in, dims_in, axis_name,
+    axis_index_groups):
+  assert prim.multiple_results
+  assert frame.name in axis_name
   if axis_index_groups is not None:
-    raise NotImplementedError("axis_index_groups not implemented in vmap collectives. "
+    raise NotImplementedError("axis_index_groups not supported in vmap collectives. "
                               "Please open a feature request!")
-  vals_out = [if_mapped(v, d) if d is not batching.not_mapped else if_unmapped(v, axis_size)
-              for v, d in zip(vals_in, dims_in)]
-  dims_out = [batching.not_mapped] * len(vals_in)
-  return vals_out, dims_out
+  vals_out = [if_mapped(v, d) if d is not batching.not_mapped
+              else if_unmapped(v, frame.size) for v, d in zip(vals_in, dims_in)]
+  if len(axis_name) > 1:
+    remaining_axis_names = tuple(n for n in axis_name if n != frame.name)
+    vals_out = prim.bind(*vals_out, axis_name=remaining_axis_names,
+                         axis_index_groups=None)
+  return vals_out, [batching.not_mapped] * len(vals_out)
+
+# TODO(mattjj): update pmin/pmax to have multiple_results like psum, delete this
+def _batched_reduction_collective2(
+    prim, if_mapped, if_unmapped, frame, vals_in, dims_in, axis_name,
+    axis_index_groups):
```

Comment:
This is missing a check that `axis_index_groups` are `None`!

Reply:
Also, couldn't we implement this in terms of a call to some list wrapping + `_batched_reduction_collective` + list unwrapping? This would make sure that it's easier to keep the impls in sync and would avoid errors like this one.

Reply:
Good points! Though #5011 will remove this function anyway so I think we don't need to refactor much on this PR. I can add the `None` check though.